feature/more/src/main/java/com/epmedu/animeal/more/profile/ProfileViewModel.kt

```diff
@@ -1,15 +1,135 @@
+                updateState { copy(surname = event.surname) }
+            }
+            is ProfileEvent.EmailChanged -> {
+                updateState { copy(email = event.email) }
+            }
+            is ProfileEvent.BirthDateChanged -> {
+                updateState {
+                    copy(
+                        birthDate = event.birthDate,
+                        formattedBirthDate = formatBirthDate(event.birthDate)
+                    )
+                }
+            }
+            ProfileEvent.Edit -> {
+                updateState {
+                    copy(readonly = false)
+                }
+            }
+            ProfileEvent.Discard -> {
+                discardChanges()
+            }
+            ProfileEvent.Save -> {
+                saveChanges()
+            }
+            ProfileEvent.ValidateName -> {
+                updateState {
+                    copy(nameError = validator.validateName(state.name).errorMessage)
+                }
+            }
+            ProfileEvent.ValidateSurname -> {
+                updateState {
+                    copy(surnameError = validator.validateSurname(state.surname).errorMessage)
+                }
+            }
+            ProfileEvent.ValidateEmail -> {
+                updateState {
+                    copy(emailError = validator.validateEmail(state.email).errorMessage)
+                }
+            }
+            ProfileEvent.ValidateBirthDate -> {
+                updateState {
+                    copy(birthDateError = validator.validateBirthDate(state.formattedBirthDate).errorMessage)
+                }
+            }
+        }
+    }
+
+    private fun discardChanges() {
+        updateState {
+            copy(readonly = true)
+        }
+        loadProfile()
+    }
+
+    private fun saveChanges() {
+        updateState {
+            copy(
+                nameError = validator.validateName(state.name).errorMessage,
+                surnameError = validator.validateSurname(state.surname).errorMessage,
+                emailError = validator.validateEmail(state.email).errorMessage,
+                birthDateError = validator.validateBirthDate(state.formattedBirthDate).errorMessage
+            )
+        }
+
+        if (state.hasErrors()) {
+            return
+        } else {
+            updateState {
+                copy(readonly = true)
+            }
+        }
+
+        val profile = Profile(
+            firstName = state.name,
+            lastName = state.surname,
+            email = state.email,
+            birthDate = state.birthDate
+        )
+
+        viewModelScope.launch {
+            profileRepository.saveProfile(profile).collect()
+        }
     }
 
-    private companion object {
-        private const val TAG = "ProfileViewModel"
+    private fun loadProfile() {
+        viewModelScope.launch {
+            profileRepository.getProfile()
+                .flowOn(Dispatchers.IO)
```

Comment:
as I remember `flowOn` is not necessary for DataStore, due to it uses this dispatcher under the hood

Reply:
Updated